### PR TITLE
Refactor shell execute functions.

### PR DESCRIPTION
**Pull request by cwize1**
This change adds a new `shell.ExecBuilder` type. This type uses the builder pattern and is used to execute sub-processes. This makes it easier to add new features to API without experiencing an exponential explosion in API size, like what can happen with the flat functions.

The `shell.ExecBuilder` type adds the ability to set the working directory and the sub-process's environment variables. This is needed for upcoming features in the Image Customizer tool.

All the existing `shell` package flat function have been rewritten to use the `shell.ExecBuilder` type. Though, some of the functions were removed in favor of using `shell.ExecBuilder` directly.

In addition, sub-process executions that capture stdout/stderr, can now also log the stdout/stderr lines. The behavior of `shell.Execute` has been changed to log stdout to TRACE and stderr to WARN.

Also, the `shell.ExecBuilder` API now makes it possible to set the log level that the stdout and stderr lines are logged to without duplicating the lines in TRACE.

---

###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] If you are adding/removing a .spec file that has multiple-versions supported, please add [@microsoft/cbl-mariner-multi-package-reviewers](https://github.com/orgs/microsoft/teams/cbl-mariner-multi-package-reviewers) team as reviewer [(Eg. golang has 2 versions 1.18, 1.21+)](https://github.com/microsoft/azurelinux/tree/2.0/SPECS/golang)
- [x] Ready to merge

###### Does this affect the toolchain?  

NO

###### Test Methodology

- Manually ran image customizer tool.

### CODE REVIEW

**Comment by dmcilvaney** [comment]
Just an fyi I have an old PR that touches some of this: https://github.com/microsoft/azurelinux/pull/6975
This PR looks like a much better solution that as you said, yet more variations of the functions.

My PR also messes with the terminate code paths for chroots. I'd still like to see that hardened a bit but it can be later.

**Review by neha170 — APPROVED** [review]
LGTM

File: toolkit/tools/internal/shell/execbuilder.go [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+	return b
+}
+
+func (b ExecBuilder) Execute() error {
+	_, _, err := b.executeHelper(false /*captureOutput*/)
+	return err
+}
+
+func (b ExecBuilder) ExecuteCaptureOuput() (string, string, error) {
+	return b.executeHelper(true /*captureOutput*/)
```
[the remark this answers is not in the document]
Comment (cwize1) [reply toolkit/tools/internal/shell/execbuilder.go:133]: `/*captureOutput*/` is a comment within Go, similar to other C-like languages.

See, https://go.dev/ref/spec#Comments

**Review by Adub17030MS — APPROVED** [review]
Lgtm!

**Pull request merged**